File: src/rbry/corelib/float.py
from types import NotImplementedType
from RBM import __JS_OP__, __JS_AS_NUMBER__, __JS_WRITE__, number

class float(number):
    def __new__(cls, o: object, /) -> float:
        if type(o) is str:
            if o == "infinity" or o == "inf":
                return __JS_WRITE__("Number.POSITIVE_INFINITY") # pyright: ignore[reportReturnType]
            if o == "-infinity" or o == "-inf":
                return __JS_WRITE__("Number.NEGATIVE_INFINITY") # pyright: ignore[reportReturnType]
            return __JS_AS_NUMBER__(o)
        return type(o).__float__(o) # type: ignore
    
    def __eq__(self, o: object, /) -> bool:
        return __JS_OP__(self, "==", o)    

    def __add__(self, o: object, /) -> NotImplementedType|float:
        match o:
            case float(): return __JS_OP__(self, "+", o)
            case int  (): return __JS_OP__(self, "+", __JS_AS_NUMBER__(o))
            case _      : return NotImplemented

    def __radd__(self, o: object, /) -> NotImplementedType|float:
        match o:
            case float(): return __JS_OP__(o                  , "+", self)
            case int  (): return __JS_OP__(__JS_AS_NUMBER__(o), "+", self)
            case _      : return NotImplemented

    def __sub__(self, o: object, /) -> NotImplementedType|float:
        match o:
            case float(): return __JS_OP__(self, "-", o)
            case int  (): return __JS_OP__(self, "-", __JS_AS_NUMBER__(o))
            case _      : return NotImplemented

    def __rsub__(self, o: object, /) -> NotImplementedType|float:
        match o:
            case float(): return __JS_OP__(o                  , "-", self)
            case int  (): return __JS_OP__(__JS_AS_NUMBER__(o), "-", self)
            case _      : return NotImplemented

    def __mul__(self, o: object, /) -> NotImplementedType|float:
        match o:
            case float(): return __JS_OP__(self, "*", o)
            case int  (): return __JS_OP__(self, "*", __JS_AS_NUMBER__(o))
            case _      : return NotImplemented

    def __rmul__(self, o: object, /) -> NotImplementedType|float:
        match o:
            case float(): return __JS_OP__(o                  , "*", self)
            case int  (): return __JS_OP__(__JS_AS_NUMBER__(o), "*", self)
            case _      : return NotImplemented

    def __div__(self, o: object, /) -> NotImplementedType|float:
        match o:
            case float(): return __JS_OP__(self, "/", o)
            case int  (): return __JS_OP__(self, "/", __JS_AS_NUMBER__(o))
            case _      : return NotImplemented

    def __rdiv__(self, o: object, /) -> NotImplementedType|float:
        match o:
            case float(): return __JS_OP__(o                  , "/", self)
            case int  (): return __JS_OP__(__JS_AS_NUMBER__(o), "/", self)
            case _      : return NotImplemented

    ### unary operators ###

    def __neg__(self, /) -> float:
        return __JS_OP__("-", self)
    
    def __int__(self, /) -> int:
        return __JS_RUN__('(x) => Math.trunc(x)', self) # type: ignore
    
    def __abs__(self, /) -> int:
         return __JS_RUN__('(x) => Math.abs(x)', self) # type: ignore

File: src/rbry/corelib-aot/float.js
;
;
var float = (() =>{
        function float() {
            return Object.create(float.prototype);
        }
    float.prototype.__new__ = function __new__(o, ){const cls = this;
        if( $RB.call(type, o, ) === str ) {
        if( $RB.op(o, "eq", "infinity")||$RB.op(o, "eq", "inf") ) {
        return Number.POSITIVE_INFINITY;

    };
if( $RB.op(o, "eq", "-infinity")||$RB.op(o, "eq", "-inf") ) {
        return Number.NEGATIVE_INFINITY;

    };
return Number(o);

    };
return $RB.mcall($RB.call(type, o, ), "__float__", o,  );

    };
float.prototype.__eq__ = function __eq__(o, ){const self = this;
        return self == o;

    };
float.prototype.__add__ = function __add__(o, ){const self = this;
        {const tname = type(o).name;if( tname === "float") {
            return self + o;

        }else if( tname === "int") {
            return self + Number(o);

        }else {
            return NotImplemented;

    }};

    };
float.prototype.__radd__ = function __radd__(o, ){const self = this;
        {const tname = type(o).name;if( tname === "float") {
            return o + self;

        }else if( tname === "int") {
            return Number(o) + self;

        }else {
            return NotImplemented;

    }};

    };
float.prototype.__sub__ = function __sub__(o, ){const self = this;
        {const tname = type(o).name;if( tname === "float") {
            return self - o;

        }else if( tname === "int") {
            return self - Number(o);

        }else {
            return NotImplemented;

    }};

    };
float.prototype.__rsub__ = function __rsub__(o, ){const self = this;
        {const tname = type(o).name;if( tname === "float") {
            return o - self;

        }else if( tname === "int") {
            return Number(o) - self;

        }else {
            return NotImplemented;

    }};

    };
float.prototype.__mul__ = function __mul__(o, ){const self = this;
        {const tname = type(o).name;if( tname === "float") {
            return self * o;

        }else if( tname === "int") {
            return self * Number(o);

        }else {
            return NotImplemented;

    }};

    };
float.prototype.__rmul__ = function __rmul__(o, ){const self = this;
        {const tname = type(o).name;if( tname === "float") {
            return o * self;

        }else if( tname === "int") {
            return Number(o) * self;

        }else {
            return NotImplemented;

    }};

    };
float.prototype.__div__ = function __div__(o, ){const self = this;
        {const tname = type(o).name;if( tname === "float") {
            return self / o;

        }else if( tname === "int") {
            return self / Number(o);

        }else {
            return NotImplemented;

    }};

    };
float.prototype.__rdiv__ = function __rdiv__(o, ){const self = this;
        {const tname = type(o).name;if( tname === "float") {
            return o / self;

        }else if( tname === "int") {
            return Number(o) / self;

        }else {
            return NotImplemented;

    }};

    };
float.prototype.__neg__ = function __neg__(){const self = this;
        return -self;

    };
float.prototype.__int__ = function __int__(){const self = this;
        return ((x) => Math.trunc(x))(self);

    };
float.prototype.__abs__ = function __abs__(){const self = this;
        return ((x) => Math.abs(x))(self);

    };

        return float;
    })();
    ;

const __exported__ = {float}